### ISSUE TITLE
Parse and Print Hexadecimal floating point literals?

### ISSUE THREAD

**Issue by NAThompson** [boostorg/multiprecision#121]
Would it be nice to have the ability to parse and print hexadecimal floating point literals in multprecision? float128 can parse as a hexadecimal, but not print (maybe a quadmath enhancement rather than Boost.Multiprecision?) but `cpp_bin_float_50` can neither parse nor print.

```cpp
#include <iostream>
#include <boost/multiprecision/cpp_bin_float.hpp>
#include <boost/multiprecision/float128.hpp>

using boost::multiprecision::cpp_bin_float_50;
using boost::multiprecision::float128;

int main()
{
    float128 x = 7;
    std::cout << std::hexfloat << x << "\n";
    cpp_bin_float_50 y = 7;

    std::cout << std::hexfloat << y << "\n";
    float128 z = 0x1p1Q;

    std::cout << "z = " << z << "\n";
    cpp_bin_float_50 w{"0x1p1"};
}
```

**Comment by jzmaddock** [comment]
It would indeed be nice, we have an open request for it somewhere I think (ah it's here: https://svn.boost.org/trac10/ticket/12069).

Fancy writing this?  ;)

**Comment by NAThompson** [comment]
I actually have been thinking about getting into the multiprecision codebase, and this looks like an achievable first task.

Lemme take a look and if I get bogged down I'll holler . . .

**Comment by ambreshbiradar9** [comment]
> Would it be nice to have the ability to parse and print hexadecimal floating point literals in multprecision? float128 can parse as a hexadecimal, but not print (maybe a quadmath enhancement rather than Boost.Multiprecision?) but `cpp_bin_float_50` can neither parse nor print.
> 
> ```c++
> #include <iostream>
> #include <boost/multiprecision/cpp_bin_float.hpp>
> #include <boost/multiprecision/float128.hpp>
> 
> using boost::multiprecision::cpp_bin_float_50;
> using boost::multiprecision::float128;
> 
> int main()
> {
>     float128 x = 7;
>     std::cout << std::hexfloat << x << "\n";
>     cpp_bin_float_50 y = 7;
> 
>     std::cout << std::hexfloat << y << "\n";
>     float128 z = 0x1p1Q;
> 
>     std::cout << "z = " << z << "\n";
>     cpp_bin_float_50 w{"0x1p1"};
> }
> ```

C:\..\..\boost/detail/iterator.hpp(13): note: This header is deprecated. Use <iterator> instead.
C:\..\..\boost/multiprecision/float128.hpp(31): fatal error C1189: #error:  "Sorry compiler is neither GCC, not Intel, don't know how to configure this header."

I use MSVC compiler

**Comment by NAThompson** [comment]
@ambreshbiradar9 : `float128` is only available with `gcc` and `intel` compiler. But I've already implemented the `float128` printing and parsing. So just remove the `float128` lines and everything should work.

**Comment by pabristow** [comment]
Correct, but if one wants to make code 'portable' over many systems, I have found this useful

Assuming that you have some Boost include above this, so that Boost.Config has run to set up the Boost macros:

`#ifdef BOOST_HAS_FLOAT128
  // Use native 128 quad (software).
  BOOST_PRAGMA_MESSAGE("Using native 128-bit quad type.") // Perhaps
  #include <boost/multiprecision/float128.hpp> // Native 128-bit float not available for MSVC or Clang (yet).
  using boost::multiprecision::float128;
#endif`


and

#ifdef BOOST_HAS_FLOAT128 // Use native 128 quad (software).
		do_something<float128>();
#endif

HTH

**Comment by pabristow** [comment]
And, yes of course it would be nice for std::hexfloat to work properly - in particular to ensure that loopback/round-tripping works.  Support is patchy, and not just for multiprecision :-(

Values are already stored as significand and exponent, so what could possibly be difficult ;-)
